File: 85.py
class Solution(object):
    def are(self,v):
        n = len(v)
        res = 0
        for i in range(0,n):
            if(i+1<n and v[i] <= v[i+1]):
                continue
            h = v[i]
            for j in range(i,-1, -1):
                h = min(h,v[j])
                res = max(h*(i-j+1),res)
        return res
                
    def maximalRectangle(self, matrix):
        row = len(matrix)
        if(row == 0):
            return 0;
        col = len(matrix[0])
        res = 0
        v = [0]*col
        for i in range(0,row):
            for j in range(0,col):
                if(matrix[i][j] == '0'):
                    v[j] = 0
                else:
                    v[j] += 1
            d = self.are(v)
            res = max(res,d)
        return res

File: 85.cpp
class Solution {
public:
    int maximalRectangle(vector<vector<char>>& matrix) {
        int row = matrix.size();
        if(row == 0) return 0;
        int col = matrix[0].size();
        std::vector<int> high(col,0);
        int res = 0;
        for(int i = 0; i < row; ++i){
            for(int j = 0; j < col; ++j){
                if(matrix[i][j] == '0'){ 
                    high[j] = 0;
                    continue;
                }
                ++high[j];
                int cnt = INT_MAX;
                int l = j;
                while(l >= 0 && high[l] != 0){
                    cnt = std::min(cnt, high[l]);
                    res = std::max(res, (j - l + 1) * cnt);
                    --l;
                }
            }
        }
        return res;
    }
};
